### PR TITLE
Move DDQ Backend functional tests to their own repo

### PR DESCRIPTION
**Pull request by quantumew**
Made some small changes after moving. Fixed some bugs and added a script to run the tests in the desired order.

### CODE REVIEW

File: README.md
Comment (fidian) [inline README.md:21]: Extra newline above headings, except the first one.

File: README.md
Comment (fidian) [inline README.md:28]: You can borrow this newline from here if you like.  That is recycling the bytes on the computer.  ♲

File: README.md
Comment (fidian) [inline README.md]: Why would any MySQL-specific things live within this repository?

File: run-tests.sh
Comment (fidian) [inline run-tests.sh]: The success/failure status code won't be returned for the first two tests.  How would a CI system catch problems?

File: script/polling-heartbeat-test.js
Comment (fidian) [inline script/polling-heartbeat-test.js]: Why do you assume the plugin is located here?

File: script/wrapped-message-tests.js
Comment (fidian) [inline script/wrapped-message-tests.js]: This is extremely MySQL specific.  I don't see how it would be possible for the MongoDB-based plugin to benefit from this in any way.

File: script/wrapped-message-tests.js
Comment (fidian) [inline script/wrapped-message-tests.js]: This doesn't return.  Also, the rest of the code isn't contained within an `else`.  This will always say the test was successful.

Comment (fidian) [reply script/wrapped-message-tests.js]: This identical problem is also elsewhere.  Please get them all.

File: script/wrapped-message-tests.js
Comment (fidian) [inline script/wrapped-message-tests.js]: How is it that this passes without an eslint directive?

I bet you don't have the stock `.eslintrc.yaml`, `.eslintignore`, and `.gitignore` files.

**Review by fidian — CHANGES_REQUESTED** [review]

File: README.md
Comment (quantumew) [reply README.md]: They don't. These are just examples using MySql to illustrate how to use it. I will make that more clear.

File: script/polling-heartbeat-test.js
Comment (quantumew) [reply script/polling-heartbeat-test.js]: I am presuming this is required by the plugin and the setup is the same as the MySql plugin. What should I do?

File: run-tests.sh
Comment (quantumew) [inline run-tests.sh]: That is an old comment it does indeed fail if any of the tests fail.

File: script/wrapped-message-tests.js
Comment (quantumew) [reply script/wrapped-message-tests.js]: Yea I noticed that. I didn't write it. Guess I will rewrite it.

File: script/wrapped-message-tests.js
Comment (quantumew) [inline script/wrapped-message-tests.js]: Indeed just copied the tests over. gitignore is there but not the linter files. Added them. I will add a directive for it. The way it was written it would just keep running runNext over and over again and would hang so I added that line instead of rewriting the test.

Comment (quantumew) [inline script/wrapped-message-tests.js]: I take that back. While I am rewriting other stuff I might as well fix this. Was mostly trying to get this over with since it isn't my code.

File: README.md
Comment (quantumew) [reply README.md:21]: Fixed

File: README.md
Comment (quantumew) [reply README.md:28]: Fixed

File: run-tests.sh
Comment (quantumew) [reply run-tests.sh]: That was an old comment. It will exit now if tests fails.

File: script/wrapped-message-tests.js
Comment (quantumew) [reply script/wrapped-message-tests.js]: Fixed

File: script/wrapped-message-tests.js
Comment (quantumew) [reply script/wrapped-message-tests.js]: I followed the correct code standard documented here and added the standard linter files http://eslint.org/docs/rules/no-process-exit

File: script/wrapped-message-tests.js
Comment (quantumew) [reply script/wrapped-message-tests.js]: Actually I am going to defer to you on this. What should I do instead? Should this test expect a certain test data that is inserted by the plugin before the tests are run?

**Comment by quantumew** [comment]
@fidian all fixes should be here 9fb00846f5f9237ee92cae3a86f0d72334f2421c. Had two questions about the correct alternatives. Like I said this was basically copied from DDQ. I fixed some bugs but I didn't write 99% of this so I will defer to you on the proper fixes.

**Comment by fidian** [comment]
Would you update your badges to match this?

```
[![npm version][npm-badge]][npm-link]
[![Build Status][travis-badge]][travis-link]
[![Dependencies][dependencies-badge]][dependencies-link]
[![Dev Dependencies][devdependencies-badge]][devdependencies-link]
```

and

```
[codecov-badge]: https://img.shields.io/codecov/c/github/tests-always-included/ddq-backend-tests/master.svg
[codecov-link]: https://codecov.io/github/tests-always-included/ddq-backend-tests?branch=master
[dependencies-badge]: https://img.shields.io/david/tests-always-included/ddq-backend-tests.svg
[dependencies-link]: https://david-dm.org/tests-always-included/ddq-backend-tests
[devdependencies-badge]: https://img.shields.io/david/dev/tests-always-included/ddq-backend-tests.svg
[devdependencies-link]: https://david-dm.org/tests-always-included/ddq-backend-tests#info=devDependencies
[npm-badge]: https://img.shields.io/npm/v/ddq-backend-tests.svg
[npm-link]: https://npmjs.org/package/ddq-backend-tests
[travis-badge]: https://img.shields.io/travis/tests-always-included/ddq-backend-tests/master.svg
[travis-link]: http://travis-ci.org/tests-always-included/ddq-backend-tests
```

File: script/wrapped-message-tests.js
Comment (fidian) [reply script/wrapped-message-tests.js]: This is a functional test, yes?  I would use the plugin and configure it to use a low heartbeat.  Then simulate what the clients would do.  Requeue is pretty easy - insert message, get message, requeue message, then listen again and get the same message.  Heartbeat failure may be a bit worse.  You could add a message, listen, then kill the heartbeat somehow (perhaps just run these things in a subprocess and kill the subprocess), then see if there are messages.  There shouldn't be.  Then wait and check again and the message should be available.

File: script/polling-heartbeat-test.js
Comment (fidian) [reply script/polling-heartbeat-test.js]: Wouldn't the current working directory be the folder of the plugin?

    Plugin = require(process.cwd());

Comment (quantumew) [reply script/polling-heartbeat-test.js]: That is much better. Fixed. 3381ac60a5429af742107958d22f07bb6c38123e

**Comment by quantumew** [comment]
@fidian squashed commits of test rewrites. I will get another reviewer for both pull requests f67fc3c40cbd82e063763f8af23e38e3e657dc25

File: .travis.yml
Comment (kyle-long) [inline .travis.yml]: https://tests-always-included.github.io/style-guide/general/ says "The last line in a file shall have a newline after it."

File: README.md
Comment (kyle-long) [inline README.md]: Its supposed be cased like "MySQL".

I care very little about this

File: README.md
Comment (kyle-long) [inline README.md]: I don't see this file.

Comment (kyle-long) [reply README.md]: Discussed and this readme is just out of date.

File: lib/index.js
Comment (kyle-long) [inline lib/index.js]: I don't think this comment is necessary.

File: lib/index.js
Comment (kyle-long) [inline lib/index.js]: This is a good comment but the context it is in makes me a bit confused. Can we move this comment to the class description maybe? Just say "All tests will be run before erroring" or add additional details to this comment like "...which is why 'errored' was assigned to false" or whatever.

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js:212]: You might want this to go above the callback call. If it were me, I would expect my callback to be executed when everything is done.

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js]: FUNCTIONAL

This feels a bit dangerous. Every Plugin we create afterwards will be called testerTwo and this config is used globally.

I suggest making a copy
```
// Note: This is only a shallow copy. This should be changed if the config contains a sub object.
configTwo = Object.assign({}, config);
configTwo.owner = "testerTwo";
```

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js]: Consider wrapping this whole if else in another if and check `data.message` once there.

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js]: This test makes an assumption that `instance` would always get the data first. the test wouldn't work the other way around. In fact I think it would hang forever if I am reading this correctly.

Perhaps we should change it to always take one action first, then the other second, regardless of instance consumes the message.

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js]: These comments are kind of pointless. May as well remove it or if you wanted to make it better you can go into more detail like "This is a test which will consume a message and then add it back into the queue".

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js]: We discussed this sort of. It might just be me but renaming these may be beneficial. I keep getting them confused in my mind with the instances functions. This doesn't just run requeue but sets up a listener, waits to get the message and then requeues it.

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js]: Already discussed but I got confused around this area. Maybe add a comment explaining that the heartbeat function sets up a listener and first "consumes" the message and THEN runs heartbeat on it so that when `remove` is called below and its listener is set up it will not get that message for the duration of the heartbeat.

I think I may have also been confused because this is exactly how you would NOT use it in a real situation. I imagine in a real situation you would set up something like.

```
setTimeout(() => {
    message.heartbeat().
}, 5000); // every five seconds.

// do work to use the message.
// Delete the message and clean up the timeout or something.
```

**Review by kyle-long — CHANGES_REQUESTED** [review]

**Comment by kyle-long** [comment]
Forgot to mention also, `run-tests.sh` still exists. I think it should be removed.

**Comment by kyle-long** [comment]
One last thing. I think we should write a test that tests the whole deduplication part. Specifically I think it should test removal of those messages after they are consumed to make sure remove actually works in this scenario (I am skeptical).

File: lib/index.js
Comment (quantumew) [reply lib/index.js]: I have to or the linter complains

File: .travis.yml
Comment (quantumew) [reply .travis.yml]: I say it shan't. Just kidding fixed. One bit commit hash will be at the bottom.

File: README.md
Comment (quantumew) [reply README.md]: Fixed

File: README.md
Comment (quantumew) [reply README.md]: Fixed

File: lib/index.js
Comment (quantumew) [reply lib/index.js]: Fixed

File: lib/tests.js
Comment (quantumew) [reply lib/tests.js]: Added detail

File: lib/tests.js
Comment (quantumew) [reply lib/tests.js]: Renamed

File: lib/tests.js
Comment (quantumew) [reply lib/tests.js]: Indeed dumb, fixed

File: lib/tests.js
Comment (quantumew) [reply lib/tests.js]: Lol totz. I must of added that later, how wonky

File: lib/tests.js
Comment (quantumew) [reply lib/tests.js]: Yea I did that because I listen on one first so it would always get it first. I guess I shouldn't rely on that. Fixed

File: lib/tests.js
Comment (quantumew) [reply lib/tests.js:212]: Fixed the 3 instances of this

File: lib/tests.js
Comment (quantumew) [reply lib/tests.js]: Added better comments

**Comment by quantumew** [comment]
@kyle-long All code review fixes here 9df13d33a8ca3c2d18165819a59e836c3bb3c3c3

File: lib/index.js
Comment (kyle-long) [inline lib/index.js]: Could you update this comment?

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js]: I don't get this message. It wouldn't have inserted a duplicate message in this scenario would it?

File: lib/tests.js
Comment (kyle-long) [inline lib/tests.js:302]: I think this test is missing one more round of trying to consume the message. We don't know if the message is actually gone at this point right?

**Review by kyle-long — CHANGES_REQUESTED** [review]

**Review by kyle-long — APPROVED** [review]

**Pull request merged**
